x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtAuthenticationToken.java

```diff
@@ -95,4 +95,18 @@ public void clearCredentials() {
     public String toString() {
         return JwtAuthenticationToken.class.getSimpleName() + "=" + this.principal;
     }
+
+    private String buildPrincipal() {
+        StringBuilder principalBuilder = new StringBuilder();
+        JWTClaimsSet jwtClaimsSet = getJWTClaimsSet();
+        principalBuilder.append(replaceLiteralSeparator(jwtClaimsSet.getIssuer())).append(PRINCIPAL_SEPARATOR);
+        principalBuilder.append(replaceLiteralSeparator(String.join(",", new TreeSet<>(jwtClaimsSet.getAudience()))))
+            .append(PRINCIPAL_SEPARATOR);
+        principalBuilder.append(replaceLiteralSeparator(jwtClaimsSet.getSubject()));
+        return principalBuilder.toString();
```

Comment:
`sub` and `aud` are both optional for access tokens. The configuration of the realm defines the fallback values..but we can't use them here else we would be back to being dependent on the realm configuration. Can we just emit all the claims (in a fixed (alpha by key) order) ? If we optimize for the indexed audit log (assuming not a specific attack) I would expect a high overlap in the tokenized values so I would not expect a huge overhead to logging more. We already have a safety valve in that you can disable the noisy audit log, and if needed could provide another sanity check value of max claims to audit.  Having the full set of claims for unsuccessful attempts feels much more useful than a small handful of cherry-picked claims which may or may not exist.  (Tangentially we currently log auth failures by default in the JWT realm, but that is an outlier I am trying to fix ... using the audit log to help troubleshoot auth failures seems reasonable and having the full claim there is needed for troubleshooting purposes)

Reply:
I've used your suggestion to use all the claims in the token principal.